### PR TITLE
C++: use field overrides in TranslatedElement and subclasses

### PR DESCRIPTION
**Pull request by rdmarsh2**
The optimizer team suggests that this is likely to improve performance; I've not benchmarked it yet.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedExpr.qll [hunk 1]
```diff
@@ -2395,25 +2327,23 @@ class TranslatedThrowValueExpr extends TranslatedThrowExpr,
...
 }
 
 /**
  * IR translation of a `throw` expression with no argument (e.g. `throw;`).
  */
 class TranslatedReThrowExpr extends TranslatedThrowExpr {
-  TranslatedReThrowExpr() {
-    throw instanceof ReThrowExpr
-  }
+  override ThrowExpr expr;
```
Comment (dave-bartolomeo) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedExpr.qll]:
```suggestion
  override ReThrowExpr expr;
```

**Comment by jbj** [comment]
131 lines of boilerplate removed. Beautiful!

**Comment by jbj** [comment]
After this change (and the other changes we merged today) I'm getting timeouts on Wireshark when evaluating `TranslatedExpr::TranslatedExpr#class#ff`. There's a Cartesian product in the DIL, but I haven't figured out why.

```
TranslatedExpr::TranslatedExpr#class#ff(TranslatedExpr::TranslatedExpr.class this,
                                        pure Expr::Expr expr) :-
  exists(pure dontcare int _, pure dontcare @location_expr _1 |
    exprs(expr, _, _1)
  ),
  (
    (
      exists(pure dontcare int _, pure dontcare @location_expr _1 |
        exprs(expr, _, _1)
      ),
      TranslatedCall::TranslatedCall#class#ff#cpe#1(this)
    );
    TranslatedExpr::TranslatedCoreExpr#class#ff(this, expr);
    TranslatedExpr::TranslatedLoad#class#ff(this, expr);
    exists(pure dontcare Expr::NewOrNewArrayExpr _ |
      TranslatedExpr::TranslatedAllocationSize#class#ff(this, _)
    )
  )
.
```

In DNF, the last case of this DIL predicate is the following CP:
```
exprs(expr, _, _1),
TranslatedExpr::TranslatedAllocationSize#class#ff(this, _2)
```

**Comment by jbj** [comment]
I think I have a workaround to remove the Cartesian product. I'll post a PR.

**Comment by rdmarsh2** [comment]
What version of the evaluator are you using? I'm not seeing the same CP, but my QL4E is a week old...

**Comment by jbj** [comment]
It's fine after I upgraded my QL4E to Jenkins build # 1651. Has the issue also gone away for you?

**Pull request merged**